pulsar-broker/src/main/java/org/apache/pulsar/compaction/Compactor.java

```diff
@@ -0,0 +1,205 @@
+                    Future<RawMessage> f = reader.readNextAsync();
+                    if (!futures.offer(f)) {
+                        RawMessage m = futures.take().get(10, TimeUnit.SECONDS);
+                        MessageId id = m.getMessageId();
+                        String key = extractKey(m);
+
+                        if (latestForKey.get(key).equals(id)) {
+                            final CompletableFuture<Void> bkf = new CompletableFuture<>();
+                            ByteBuf serialized = m.serialize();
+                            lh.asyncAddEntry(serialized,
+                                             (rc, ledger, eid, ctx) -> {
+                                                 if (rc != BKException.Code.OK) {
+                                                     bkf.completeExceptionally(BKException.create(rc));
+                                                 } else {
+                                                     bkf.complete(null);
+                                                 }
+                                             }, null);
+                            serialized.release();
+                            lastAddFuture = bkf;
+                        }
+
+                        if (id.equals(latestMessageId)) {
+                            cancelOrFree(f);
+                            break;
+                        } else {
+                            futures.offer(f);
+                        }
+                    }
+                }
+                futures.forEach(Compactor::cancelOrFree);
+                futures.clear();
+
+                lastAddFuture.get(); // ensure all adds have completed
+                lh.close();
+
+                reader.acknowledgeCumulativeAsync(latestMessageId,
+                                                  ImmutableMap.of(COMPACTED_TOPIC_LEDGER_PROPERTY,
+                                                                  lh.getId())).get();
+                future.complete(lh.getId());
+            } catch (Exception e) {
+                future.completeExceptionally(e);
+            } finally {
+                if (reader != null) {
+                    reader.closeAsync();
+                }
+            }
+        }
+    }
+
+    private static void cancelOrFree(Future<RawMessage> f) {
+        if (!f.cancel(false)) {
+            try {
+                f.get().close();
+            } catch (InterruptedException ie) {
+                Thread.currentThread().interrupt();
+            } catch (ExecutionException e) {
+                // we don't care, we're trying to cancel anyhow
+            }
+        }
+    }
+
+    private static String extractKey(RawMessage m) throws Exception {
+        ByteBuf headersAndPayload = m.getHeadersAndPayload();
+        MessageMetadata msgMetadata = Commands.parseMessageMetadata(headersAndPayload);
+        return msgMetadata.getPartitionKey();
```

Comment:
If key would be always partition-key then does it mean that compaction keeps only 1 latest message for each individual partitioned topic?

Reply:
yes. Perhaps this could be extended in the future to allow a user to specify how to extract the key from a custom data type, but would depend on the schema stuff that's also happening now.